pkg/ottl/contexts/internal/ctxcontext/context.go

```diff
@@ -0,0 +1,306 @@
+				return nil, nil
+			}
+			return getIndexableValueFromStringArr(ctx, tCtx, keys, mdVal)
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.grpc.metadata")
+		},
+	}
+}
+
+func accessClientAuth[K any](path ottl.Path[K]) (ottl.GetSetter[K], error) {
+	nextPath := path.Next()
+	if nextPath == nil {
+		return nil, ctxerror.New(path.Name(), path.String(), Name, DocRef)
+	}
+	switch nextPath.Name() {
+	case "attributes":
+		if nextPath.Keys() == nil {
+			return accessClientAuthAttributesKeys[K](), nil
+		}
+		return accessClientAuthAttributesKey[K](nextPath.Keys()), nil
+	default:
+		return nil, ctxerror.New(nextPath.Name(), nextPath.String(), Name, DocRef)
+	}
+}
+
+func getAuthAttributeValue(attr any) (any, error) {
+	switch a := attr.(type) {
+	case string:
+		return a, nil
+	case nil:
+		return nil, nil
+	default:
+		b, err := json.Marshal(attr)
+		if err != nil {
+			return "", err
+		}
+		return string(b), nil
+	}
+}
+
+func convertAuthDataToMap(authData client.AuthData) (pcommon.Map, error) {
+	authMap := pcommon.NewMap()
+	if authData == nil {
+		return authMap, nil
+	}
+	names := authData.GetAttributeNames()
+	for _, name := range names {
+		attrVal := authData.GetAttribute(name)
+		attrStr, err := getAuthAttributeValue(attrVal)
+		if err != nil {
+			return pcommon.NewMap(), err
+		}
+		switch v := attrStr.(type) {
+		case string:
+			authMap.PutStr(name, v)
+		default:
+			authMap.PutEmpty(name)
+		}
+	}
+	return authMap, nil
+}
+
+func accessClientAuthAttributesKeys[K any]() ottl.StandardGetSetter[K] {
+	return ottl.StandardGetSetter[K]{
+		Getter: func(ctx context.Context, _ K) (any, error) {
+			cl := client.FromContext(ctx)
+			return convertAuthDataToMap(cl.Auth)
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.client.auth.attributes")
+		},
+	}
+}
+
+func accessClientAuthAttributesKey[K any](keys []ottl.Key[K]) ottl.StandardGetSetter[K] {
+	return ottl.StandardGetSetter[K]{
+		Getter: func(ctx context.Context, tCtx K) (any, error) {
+			if len(keys) == 0 {
+				return nil, errors.New("cannot get map value without keys")
+			}
+			cl := client.FromContext(ctx)
+			key, err := ctxutil.GetMapKeyName(ctx, tCtx, keys[0])
+			if err != nil {
+				return nil, err
+			}
+			if cl.Auth == nil {
+				return nil, nil
+			}
+			attrVal := cl.Auth.GetAttribute(*key)
+			attrStr, err := getAuthAttributeValue(attrVal)
+			if err != nil {
+				return nil, err
+			}
+			return attrStr, nil
+		},
+		Setter: func(_ context.Context, _ K, _ any) error {
+			return errors.New("cannot set value in context.client.auth.attributes")
+		},
+	}
+}
+
+func convertClientMetadataToMap(md client.Metadata) pcommon.Map {
+	mdMap := pcommon.NewMap()
+	for k := range md.Keys() {
+		sl := mdMap.PutEmptySlice(k)
+		mdVal := md.Get(k)
+		mdSl := convertStringArrToValueSlice(mdVal)
+		mdSl.Slice().CopyTo(sl)
+	}
+	return mdMap
```

Comment:
```suggestion
	mdMap := pcommon.NewMap()
	for k := range md.Keys() {
		convertStringArrToValueSlice(md.Get(k)).MoveTo(mdMap.PutEmpty(k))
	}
	return mdMap
```

Let's `MoveTo` instead of `CopyTo`, as the value source is only used for this function .

Reply:
done